File: src/commonMain/kotlin/me/kolterdyx/compiler/statement/VariableDeclarationStatement.kt
package me.kolterdyx.compiler.statement

import me.kolterdyx.compiler.Token
import me.kolterdyx.compiler.ast.StatementVisitor
import me.kolterdyx.compiler.expression.Expression

class VariableDeclarationStatement(
    val name: Token,
    val type: Token,
    val initializer: Expression?
) : Statement() {
    override fun <R> accept(visitor: StatementVisitor<R>): R {
        return visitor.visitVariableDeclaration(this)
    }
}

File: internal/ast/variableDeclarationStatement.go
package ast

import (
	"github.com/Kolterdyx/mcbasic/internal/interfaces"
	"github.com/Kolterdyx/mcbasic/internal/tokens"
	"github.com/Kolterdyx/mcbasic/internal/types"
)

type VariableDeclarationStmt struct {
	Name        tokens.Token
	Initializer Expr
	ValueType   types.ValueType
}

func (v VariableDeclarationStmt) Accept(visitor StatementVisitor) any {
	return visitor.VisitVariableDeclaration(v)
}

func (v VariableDeclarationStmt) Type() NodeType {
	return VariableDeclarationStatement
}

func (v VariableDeclarationStmt) ToString() string {
	if v.Initializer != nil {
		return v.Name.Lexeme + " = " + v.Initializer.ToString() + ";"
	}
	return v.Name.Lexeme + ";"
}

func (v VariableDeclarationStmt) GetSourceLocation() interfaces.SourceLocation {
	return v.Name.SourceLocation
}
